### PR TITLE
Implement title_format

### PR DESCRIPTION
**Pull request by RyanDwyer**
This implements the `title_format` command, with a new placeholder `%shell` which gets substituted with the view type (xwayland, xdg_shell_v6 or wl_shell).

Example config:

    for_window [title=".*"] title_format %title (class=%class instance=%instance shell=%shell)

You can change the title using swaymsg:

    $ swaymsg title_format %title

Some implementation notes:
* container_update_title() has been renamed to view_update_title() and moved into view.c as it only applies to views.
* Because of the above, container_notify_child_title_changed() is now public.
* The arguments of view_update_title() have changed. It used to have the title given to it via an argument; now it reads it from view_get_title().
* Containers now have both a `name` and `formatted_title` property. The `name` is used purely for `strcmp()`, so we don't blindly recreate the title textures if the view hasn't changed its title.

I've also tested it will null titles (Firefox developer tools as per #1912).

### CODE REVIEW

File: sway/commands/title_format.c [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	if (container->type != C_VIEW) {
+		return cmd_results_new(CMD_INVALID, "title_format",
+				"Only views can have a title_format");
+	}
+	struct sway_view *view = container->sway_view;
+	char *format = join_args(argv, argc);
+	if (view->title_format) {
+		free(view->title_format);
+	}
+	view->title_format = strdup(format);
```
Comment (taiyu-len) [inline sway/commands/title_format.c]: dont need the strdup here,
join_args already allocates memory for format, which is also is getting leaked here.

Comment (RyanDwyer) [reply sway/commands/title_format.c]: Thanks!

File: sway/tree/view.c [hunk 2]
```diff
@@ -475,3 +493,127 @@ void view_child_destroy(struct sway_view_child *child) {
...
+
+	// Now build the title
+	format = view->title_format;
+	next = strchr(format, '%');
+	while (next) {
+		// Copy everything up to the %
+		strncat(buffer, format, next - format);
+		format = next;
+
+		if (strncmp(next, "%title", 6) == 0) {
```
Comment (emersion) [inline sway/tree/view.c]: Is it possible to factor this into a function that returns the formatted length and only calls `strcat` if the `buffer` parameter isn't NULL? We should be able to call it first with `buffer` set to NULL, then allocate the buffer, and then call it again. That would de-duplicate this code.

File: sway/tree/view.c [hunk 3]
```diff
@@ -475,3 +493,127 @@ void view_child_destroy(struct sway_view_child *child) {
...
+		if (title && view->swayc->name && strcmp(title, view->swayc->name) == 0) {
+			return;
+		}
+		if (!title && !view->swayc->name) {
+			return;
+		}
+	}
+
+	if (title) {
+		if (view->swayc->name) {
```
Comment (emersion) [inline sway/tree/view.c]: No need for these `if` statements: `free(NULL)` is a no-op.

Comment (emersion) [reply sway/tree/view.c]: You can factor them out of the outer `if`

File: sway/commands/title_format.c [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	}
+	struct sway_view *view = container->sway_view;
+	char *format = join_args(argv, argc);
+	if (view->title_format) {
+		free(view->title_format);
+	}
+	view->title_format = format;
+	view_update_title(view, true);
+	config_find_font_height(true);
+	arrange_root();
```
Comment (emersion) [inline sway/commands/title_format.c]: Is `arrange_root` necessary here?

File: sway/tree/view.c [hunk 5]
```diff
@@ -475,3 +493,127 @@ void view_child_destroy(struct sway_view_child *child) {
 		free(child);
 	}
 }
+
+static char *parse_title_format(struct sway_view *view) {
+	if (!view->title_format || strcmp(view->title_format, "%title") == 0) {
+		return strdup(view_get_title(view));
```
Comment (emersion) [inline sway/tree/view.c]: `strdup` will segfault if `view_get_title` returns NULL

Comment (RyanDwyer) [reply sway/tree/view.c]: `parse_title_format` is only called if the title is not NULL. Should I still put a check in here?

File: include/sway/tree/container.h [hunk 6]
```diff
@@ -63,7 +63,8 @@ struct sway_container {
 	 */
 	size_t id;
 
-	char *name;
+	char *name;            // The view's title (unformatted)
```
Comment (emersion) [inline include/sway/tree/container.h]: Nitpick: maybe this should be renamed to `title` for consistency?

Comment (RyanDwyer) [reply include/sway/tree/container.h]: I kind of agree, but that also means we'd be changing `output->name` to `output->title` and `workspace->name` to `workspace->title`, and I feel like `name` is a better name for those.

Comment (emersion) [reply include/sway/tree/container.h]: Yes. you're right.

**Comment by SirCmpwn** [comment]
Thanks!
